### PR TITLE
TAU service

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
Hey, this is a template we'll be able to work with for the TAU service. We need to decouple the Regex in the RegexFilter from the service it's used in, but otherwise it seems rather solid. Ideally, we'd get

1) a build of TAU based on the Caliper services we have enabled
2) stripping the "=true" from annotations
3) Automatic generation of regex's to avoid non-stack annotations

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
calitau

The background may show up as black, that's an svg issue, but this is a screenshot of cali-basic profile in paraprof with a regex filtering the problematic annotations out.

Also might interest @khuck

**Comment by davidbeckingsale** [comment]
This looks cool. I'm not sure about how you've set up the use of filters though - is this still waiting on the runtime filter selection?

**Comment by DavidPoliakoff** [comment]
This actually has runtime filter selection working :)

File: src/services/common/ToolWrapper.h [hunk 1]
```diff
@@ -7,20 +7,53 @@
...
+    virtual std::string service_tag(){
+        throw std::runtime_error("Caliper services must implement service_tag");
+    }
     virtual void beginCallback(Caliper* c, const Attribute& attr, const Variant& value){
-        if(FilterType::filter(attr, value)){
+        if(filter->filter(attr, value)){
                 this->beginAction(c,attr,value);
         }
     }
+    Filter* getFilter(){
```
Comment (davidbeckingsale) [inline src/services/common/ToolWrapper.h]: This is the runtime filter stuff?

File: src/services/common/ToolWrapper.h [hunk 2]
```diff
@@ -7,20 +7,53 @@
...
+        if(config.get("regex").to_string()==""){
+           filter = new DefaultFilter();
+        }
+        filter = new RegexFilter();
+        filter->configure(config);
+        return filter;
+    }
+    virtual void pre_initialize(){
+        std::string name = service_tag();
+        config = cali::RuntimeConfig::init("tau",s_configdata);
```
Comment (davidbeckingsale) [inline src/services/common/ToolWrapper.h]: Why is tau mentioned here?

Comment (DavidPoliakoff) [inline src/services/common/ToolWrapper.h]: Hahaoops. Fixing

File: src/services/common/ToolWrapper.h [hunk 3]
```diff
@@ -49,7 +49,7 @@ class ToolWrapper {
     }
     virtual void pre_initialize(){
         std::string name = service_tag();
-        config = cali::RuntimeConfig::init("tau",s_configdata);
+        config = cali::RuntimeConfig::init(service_tag,s_configdata);
```
Comment (davidbeckingsale) [inline src/services/common/ToolWrapper.h]: 👍 awesome!

File: src/services/common/ToolWrapper.h [hunk 4]
```diff
@@ -7,20 +7,53 @@
...
-        if(FilterType::filter(attr, value)){
+        if(filter->filter(attr, value)){
                 this->beginAction(c,attr,value);
         }
     }
+    Filter* getFilter(){
+        if(config.get("regex").to_string()==""){
+           filter = new DefaultFilter();
+        }
+        filter = new RegexFilter();
```
Comment (daboehme) [inline src/services/common/ToolWrapper.h]: Missing an`else` ?!

File: src/services/common/ToolWrapper.h [hunk 5]
```diff
@@ -7,20 +7,53 @@
...
+    "Whether the regular expression should include or exclude annotations",
+    "Whether the regular expression should include or exclude annotations"
+  },
+  cali::ConfigSet::Terminator
+};
+
 class ToolWrapper {
+    protected:
+    cali::ConfigSet config;
+    Filter* filter;
```
Comment (daboehme) [inline src/services/common/ToolWrapper.h]: Would be good to `delete` the filter at some point

Comment (DavidPoliakoff) [inline src/services/common/ToolWrapper.h]: Discussed in person, for posterity: delete happens in ToolWrapper.h

**Pull request merged**